adapters/insticator/insticator.go

```diff
@@ -0,0 +1,328 @@
+package insticator
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v3/adapters"
+	"github.com/prebid/prebid-server/v3/config"
+	"github.com/prebid/prebid-server/v3/errortypes"
+	"github.com/prebid/prebid-server/v3/openrtb_ext"
+	"github.com/prebid/prebid-server/v3/util/jsonutil"
+	"github.com/prebid/prebid-server/v3/util/mathutil"
+)
+
+type ext struct {
+	Insticator impInsticatorExt `json:"insticator"`
+}
+
+type impInsticatorExt struct {
+	AdUnitId    string `json:"adUnitId"`
+	PublisherId string `json:"publisherId"`
+}
+
+type adapter struct {
+	endpoint string
+}
+
+type reqExt struct {
+	Insticator *reqInsticatorExt `json:"insticator,omitempty"`
+}
+
+type reqInsticatorExt struct {
+	Caller []insticatorCaller `json:"caller,omitempty"`
+}
+
+type insticatorCaller struct {
+	Name    string `json:"name,omitempty"`
+	Version string `json:"version,omitempty"`
+}
+
+// caller Info used to track Prebid Server
+// as one of the hops in the request to exchange
+var caller = insticatorCaller{"Prebid-Server", "n/a"}
+
+type bidExt struct {
+	Insticator bidInsticatorExt `json:"insticator,omitempty"`
+}
+
+type bidInsticatorExt struct {
+	MediaType string `json:"mediaType,omitempty"`
+}
+
+// Builder builds a new instance of the Insticator adapter with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// getMediaTypeForBid figures out which media type this bid is for
+func getMediaTypeForBid(bid *openrtb2.Bid) openrtb_ext.BidType {
+	switch bid.MType {
+	case openrtb2.MarkupBanner:
+		return openrtb_ext.BidTypeBanner
+	case openrtb2.MarkupVideo:
+		return openrtb_ext.BidTypeVideo
+	default:
+		return openrtb_ext.BidTypeBanner
+	}
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var adapterRequests []*adapters.RequestData
+	var groupedImps = make(map[string][]openrtb2.Imp)
+
+	reqExt, err := makeReqExt(request)
+	if err != nil {
+		errs = append(errs, err)
+	}
+
+	request.Ext = reqExt
+
+	// Create a copy of the request to avoid modifying the original request
+	requestCopy := *request
+
+	for i := 0; i < len(request.Imp); i++ {
+		impCopy, impKey, err := makeImps(request.Imp[i])
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		// Populate site.publisher.id from imp extension
+		if requestCopy.Site != nil || requestCopy.App != nil {
+			if err := populatePublisherId(&impCopy, &requestCopy); err != nil {
```

Comment:
This will unmarshal the `imp.ext` of the current imp and update the `site`/`app` object's publisher ID to match. It will repeat this process for every imp in the request. It would be much more efficient to remove this step from the loop, and just execute this for the last `imp` in the list, as it will have the same result and only consume `1/len(Imp)` as many cycles.

Reply:
Yes, I know. I was doing the same previously but someone previously asked to update it.

https://github.com/prebid/prebid-server/pull/3806#discussion_r1705391591

@hhhjort